File: interpreter/lexer.py
from typing import Union, Optional, List
from enum import Enum, auto
from interpreter.tokenizer import Token, TokenTypes
from interpreter.exceptions import UnknownSymbolError


res_keywords_toks = {
    "function": Token(TokenTypes.FUNCTION),
    "return": Token(TokenTypes.RETURN),
    "if": Token(TokenTypes.IF),
    "else": Token(TokenTypes.ELSE),
}

whitespaces_no_nl_toks = (
    " ",
    "\t",
    "\r",
    "\f",
    "\v",
)

kw_with_trailing_ws = ("function", "return")
kw_with_trailing_lparen = ("if", "return")

single_chr_toks = {
    "+": Token(TokenTypes.PLUS),
    "-": Token(TokenTypes.MINUS),
    "*": Token(TokenTypes.MUL),
    "/": Token(TokenTypes.DIV),
    "=": Token(TokenTypes.ASSIGN),
    "<": Token(TokenTypes.LT),
    ">": Token(TokenTypes.GT),
    "/": Token(TokenTypes.DIV),
    "(": Token(TokenTypes.LPAREN),
    ")": Token(TokenTypes.RPAREN),
    "{": Token(TokenTypes.LBRACE),
    "}": Token(TokenTypes.RBRACE),
    ",": Token(TokenTypes.COMMA),
    ";": Token(TokenTypes.SEMI),
    "\n": Token(TokenTypes.EOL),
    "[": Token(TokenTypes.LBRACE),
    "]": Token(TokenTypes.RBRACE),
}

mult_chr_toks = {
    "==": Token(TokenTypes.EQ),
    "<=": Token(TokenTypes.LE),
    ">=": Token(TokenTypes.GE),
}


class Lexer:
    def __init__(self, input: str = ""):
        self._buffer = input
        self._pos = 0

    @property
    def buffer(self):
        return self._buffer

    @property
    def pos(self):
        return self._pos

    @property
    def current_char(self) -> Optional[str]:
        if self._pos >= len(self._buffer):
            return None
        return self._buffer[self._pos]

    def _advance(self, offset=1) -> None:
        if offset < 0:
            raise IndexError(f"{self._advance.__name__} can only advance forward")
        pos = self._pos
        if pos + offset > len(self._buffer):
            raise IndexError("Index out of range")
        self._pos += offset

    def _peek(self, offset=1) -> Optional[str]:
        if offset < 0:
            raise IndexError(f"{self._peek.__name__}can only peek forward")
        pos = self._pos
        if pos + offset >= len(self._buffer):
            return None
        return self._buffer[pos + offset]

    def _look_ahead(self, offset=1) -> Optional[str]:
        if offset < 0:
            raise IndexError(f"{self._look_ahead.__name__}can only peek forward")
        pos = self._pos
        if pos + offset > len(self._buffer):
            return None
        return self._buffer[pos : pos + offset]

    def _is_eof(self) -> Optional[Token]:
        if not self._buffer:
            return Token(TokenTypes.EOF)
        if self._pos >= len(self._buffer):
            return Token(TokenTypes.EOF)
        return None

    def _num(self) -> Optional[Token]:
        s = ""
        curr_char = self.current_char
        while curr_char and curr_char.isdigit():
            s += curr_char
            self._advance()
            curr_char = self.current_char
        return Token(TokenTypes.INTEGER, int(s))

    def _id(self) -> Optional[Token]:
        s = ""
        curr_char = self.current_char
        while curr_char and curr_char.isalnum():
            s += curr_char
            self._advance()
            curr_char = self.current_char
        return Token(TokenTypes.ID, s)

    def _string(self):
        self._advance()
        s = ""
        while self.current_char and self.current_char != '"':
            s += self.current_char
            self._advance()
        self._advance()
        return Token(TokenTypes.STRING, s)

    def _get_next_token(self) -> Optional[Token]:
        eof = self._is_eof()
        if eof:
            return eof

        buffer, pos = self._buffer, self._pos
        curr_char = self.current_char
        for tok in res_keywords_toks:
            if self._look_ahead(len(tok)) == tok:
                # TODO: check if it's robust enough
                following_chr = self._peek(len(tok))
                if following_chr and following_chr.isalnum():
                    break
                # if tok in kw_with_trailing_lparen and self._peek(len(tok)) not in whitespaces_no_nl_toks
                not_res_kw = False
                print(tok)
                for tok2 in res_keywords_toks:
                    if tok == tok2:
                        continue
                    if self._look_ahead(len(tok) + len(tok2)) == tok + tok2:
                        not_res_kw = True
                if not_res_kw:
                    break
                self._advance(len(tok))
                return res_keywords_toks[tok]

        for tok in mult_chr_toks:
            if self._look_ahead(len(tok)) == tok:
                self._advance(len(tok))
                return mult_chr_toks[tok]

        if curr_char == '"':
            return self._string()

        elif curr_char in single_chr_toks.keys():
            self._advance()
            return single_chr_toks[curr_char]

        elif curr_char and curr_char.isdigit():
            return self._num()

        elif curr_char and curr_char.isalnum():
            return self._id()

        else:
            raise UnknownSymbolError({buffer[pos:].split()[0]})

        return None

    def _skip_spaces(self) -> None:
        input_len = len(self._buffer)
        while self._pos < input_len and self._buffer[self._pos] in whitespaces_no_nl_toks:
            self._pos += 1

    def get_tokens(self) -> List[Token]:
        tokens = []
        self._skip_spaces()
        curr_token = self._get_next_token()
        while curr_token and curr_token._type != TokenTypes.EOF:
            self._skip_spaces()
            tokens.append(curr_token)
            curr_token = self._get_next_token()
        tokens.append(Token(TokenTypes.EOF))
        return tokens

File: src/lexer.zig
const std = @import("std");
const dbg = @import("./debug.zig");
const Token = @import("./tokens.zig").Token;
const TokenType = @import("./tokens.zig").TokenType;
const reserved = @import("./lex_constants.zig").reserved;
const op_math = @import("./lex_constants.zig").math;
const op_cmp = @import("./lex_constants.zig").cmp;
const op_assign = @import("./lex_constants.zig").assign;
const delimeters = @import("./lex_constants.zig").delimeters;
const separators = @import("./lex_constants.zig").separators;

pub const Error = error{ NullSource, NullTokens, BadToken, OutOfBounds, UnterminatedString };

const reserved_kws = std.StaticStringMap(TokenType).initComptime(.{
    .{ reserved._function, TokenType.function_kw },
    .{ reserved._return, TokenType.return_kw },
    .{ reserved._if, TokenType.if_kw },
    .{ reserved._else, TokenType.else_kw },
    .{ reserved._while, TokenType.while_kw },
    .{ reserved._for, TokenType.for_kw },
    .{ reserved._break, TokenType.break_kw },
    .{ reserved._continue, TokenType.continue_kw },
});

const single_chr_toks = std.StaticStringMap(TokenType).initComptime(.{
    .{ op_math.plus, TokenType.plus },
    .{ op_math.minus, TokenType.minus },
    .{ op_math.mul, TokenType.mul },
    .{ op_math.div, TokenType.div },
    .{ op_math.mod, TokenType.mod },
    .{ op_cmp.lt, TokenType.lt },
    .{ op_assign.assign, TokenType.assign },
    .{ op_cmp.gt, TokenType.gt },
    .{ delimeters.lparen, TokenType.lparen },
    .{ delimeters.rparen, TokenType.rparen },
    .{ delimeters.lbrace, TokenType.lbrace },
    .{ delimeters.rbrace, TokenType.rbrace },
    .{ delimeters.lbrack, TokenType.lbrack },
    .{ delimeters.rbrack, TokenType.rbrack },
    .{ separators.comma, TokenType.comma },
    .{ separators.semi, TokenType.semi },
    .{ "\n", TokenType.eol },
});

const mult_chr_toks = std.StaticStringMap(TokenType).initComptime(.{
    .{ "==", TokenType.eq },
    .{ "<=", TokenType.le },
    .{ ">=", TokenType.ge },
});

const whitespaces_no_nl = std.StaticStringMap(undefined).initComptime(.{
    .{" "},
    .{"\t"},
    .{"\r"},
});

pub const Lexer: type = struct {
    const Self = @This();
    allocator: std.mem.Allocator = undefined,
    source: ?([]const u8) = undefined,
    pos: usize = 0,
    line: usize = 0,
    tokens: ?std.ArrayList(Token),

    pub fn init(buffer: []const u8, allocator: std.mem.Allocator) !Self {
        const tokens = std.ArrayList(Token).init(allocator);
        const lexer = Self{ .allocator = allocator, .source = buffer, .tokens = tokens };
        return lexer;
    }

    pub fn deinit(self: *Self) void {
        for (self.tokens.?.items) |*tok| {
            tok.deinit();
            tok.lexeme = null;
        }
        self.tokens.?.deinit();
        self.tokens = null;
    }

    pub fn tokenize(self: *Self) !void {
        if (self.tokens) |*tokens| {
            var next_tok = try self.nextToken();
            try tokens.append(next_tok);
            while (next_tok.type != TokenType.eof) {
                dbg.print("{}: '{s}' L:{}\n", .{ next_tok.type, next_tok.lexeme.?, next_tok.line }, @src());
                next_tok = try self.nextToken();
                try tokens.*.append(next_tok);
            }
        } else {
            return Error.NullTokens;
        }
    }

    pub fn getTokens(self: *const Self) !std.ArrayList(Token) {
        return self.tokens orelse Error.NullTokens;
    }

    pub fn nextToken(self: *Self) !Token {
        self.skipWhitespace() catch {
            return Token.init(TokenType.eof, "", self.line, self.allocator);
        };

        if (self.isAtEnd()) {
            return Token.init(TokenType.eof, "", self.line, self.allocator);
        }

        for (reserved_kws.keys()) |kw| {
            if (std.mem.eql(u8, self.lookAhead(kw.len), kw)) {
                const following_chr = self.peek(kw.len);
                if (following_chr != 0x0 and std.ascii.isAlphanumeric(following_chr)) {
                    continue;
                }
                try self.advance(kw.len);
                const tokType = reserved_kws.get(kw) orelse unreachable;
                return Token.init(tokType, kw, self.line, self.allocator);
            }
        }

        for (mult_chr_toks.keys()) |kw| {
            if (std.mem.eql(u8, self.lookAhead(kw.len), kw)) {
                try self.advance(kw.len);
                const tokType = mult_chr_toks.get(kw) orelse unreachable;
                return Token.init(tokType, kw, self.line, self.allocator);
            }
        }

        if (self.peek(0) == '"') {
            return try self.string();
        }

        for (single_chr_toks.keys()) |kw| {
            const single_chr = kw[0];
            if (self.peek(0) == single_chr) {
                try self.advance(1);
                const tokType = single_chr_toks.get(kw) orelse unreachable;
                return Token.init(tokType, kw, self.line, self.allocator);
            }
        }

        if (std.ascii.isDigit(self.peek(0))) {
            return try self.num();
        }

        if (std.ascii.isAlphabetic(self.peek(0))) {
            return try self.id();
        }

        try self.advance(1);
        return Error.BadToken;
    }

    // Helper functions
    fn string(self: *Self) !Token {
        const start = self.pos;
        try self.advance(1);
        while (self.peek(0) != '"' and self.peek(0) != 0x00) {
            try self.advance(1);
        }
        if (self.pos == self.source.?.len) {
            return Error.UnterminatedString;
        }
        try self.advance(1);
        return Token.init(TokenType.string, self.source.?[start..self.pos], self.line, self.allocator);
    }

    fn num(self: *Self) !Token {
        const start = self.pos;
        var n: i28 = 0;
        while (std.ascii.isDigit(self.peek(0))) {
            n *= 10;
            n += @intCast(self.peek(0));
            try self.advance(1);
        }
        return Token.init(TokenType.integer, self.source.?[start..self.pos], self.line, self.allocator);
    }

    fn id(self: *Self) !Token {
        const start = self.pos;
        while (std.ascii.isAlphanumeric(self.peek(0))) {
            try self.advance(1);
        }
        return Token.init(TokenType.id, self.source.?[start..self.pos], self.line, self.allocator);
    }

    // Utils functions
    inline fn peek(self: *const Self, offset: ?usize) u8 {
        const offs = offset orelse 0;
        return if (self.pos + offs < self.source.?.len) self.source.?[self.pos + offs] else 0x00;
    }

    inline fn lookAhead(self: *const Self, len: usize) []const u8 {
        if (self.pos + len >= self.source.?.len) {
            return "";
        }
        return self.source.?[self.pos .. self.pos + len];
    }

    inline fn advance(self: *Self, offset: ?usize) !void {
        const offs = offset orelse 1;
        if (self.pos + offs <= self.source.?.len) {
            self.pos += offs;
        } else {
            return Error.OutOfBounds;
        }
    }

    inline fn isAtEnd(self: *const Self) bool {
        return self.pos >= (self.source orelse unreachable).len;
    }

    fn skipWhitespace(self: *Self) Error!void {
        while (!self.isAtEnd()) {
            switch (self.peek(0)) {
                ' ', '\r', '\t' => try self.advance(1),
                '\n' => {
                    self.line += 1;
                    try self.advance(1);
                },
                else => break,
            }
        } else {
            return Error.OutOfBounds;
        }
    }
};
